File: Sources/FWObjC/Module/Controller/ScrollViewController.m
//
//  ScrollViewController.m
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

#import "ScrollViewController.h"
#import <objc/runtime.h>

#if FWMacroSPM

@interface UIView ()

- (NSArray<NSLayoutConstraint *> *)fw_pinEdgesToSuperview:(UIEdgeInsets)insets;

@end

#else

#import <FWFramework/FWFramework-Swift.h>

#endif

#pragma mark - __FWViewControllerManager+__FWScrollViewController

@implementation __FWViewControllerManager (__FWScrollViewController)

+ (void)load
{
    __FWViewControllerIntercepter *intercepter = [[__FWViewControllerIntercepter alloc] init];
    intercepter.viewDidLoadIntercepter = @selector(scrollViewControllerViewDidLoad:);
    intercepter.forwardSelectors = @{
        @"scrollView" : @"fw_innerScrollView",
        @"contentView" : @"fw_innerContentView",
        @"setupScrollLayout" : @"fw_innerSetupScrollLayout",
    };
    [[__FWViewControllerManager sharedInstance] registerProtocol:@protocol(__FWScrollViewController) withIntercepter:intercepter];
}

- (void)scrollViewControllerViewDidLoad:(UIViewController<__FWScrollViewController> *)viewController
{
    UIScrollView *scrollView = [viewController scrollView];
    [viewController.view addSubview:scrollView];
    
    UIView *contentView = [viewController contentView];
    [scrollView addSubview:contentView];
    [contentView fw_pinEdgesToSuperview:UIEdgeInsetsZero];
    
    if (self.hookScrollViewController) {
        self.hookScrollViewController(viewController);
    }
    
    if ([viewController respondsToSelector:@selector(setupScrollView)]) {
        [viewController setupScrollView];
    }
    
    [viewController setupScrollLayout];
    [scrollView setNeedsLayout];
    [scrollView layoutIfNeeded];
}

@end

#pragma mark - UIViewController+__FWScrollViewController

@interface UIViewController (__FWScrollViewController)

@end

@implementation UIViewController (__FWScrollViewController)

- (UIScrollView *)fw_innerScrollView
{
    UIScrollView *scrollView = objc_getAssociatedObject(self, _cmd);
    if (!scrollView) {
        scrollView = [[UIScrollView alloc] init];
        scrollView.showsVerticalScrollIndicator = NO;
        scrollView.showsHorizontalScrollIndicator = NO;
        objc_setAssociatedObject(self, _cmd, scrollView, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    }
    return scrollView;
}

- (UIView *)fw_innerContentView
{
    UIView *contentView = objc_getAssociatedObject(self, _cmd);
    if (!contentView) {
        contentView = [[UIView alloc] init];
        objc_setAssociatedObject(self, _cmd, contentView, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    }
    return contentView;
}

- (void)fw_innerSetupScrollLayout
{
    UIScrollView *scrollView = [(id<__FWScrollViewController>)self scrollView];
    [scrollView fw_pinEdgesToSuperview:UIEdgeInsetsZero];
}

@end

File: Sources/FWUIKit/Module/Controller/ScrollViewController.swift
//
//  ScrollViewController.swift
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

import UIKit
#if FWMacroSPM
@_spi(FW) import FWFramework
#endif

// MARK: - ScrollViewControllerProtocol
/// 滚动视图控制器协议，可覆写
@MainActor public protocol ScrollViewControllerProtocol: ViewControllerProtocol {
    /// 滚动视图，默认不显示滚动条
    var scrollView: UIScrollView { get }

    /// 内容容器视图，自动撑开，子视图需要添加到此视图上
    var contentView: UIView { get }

    /// 渲染滚动视图，setupSubviews之前调用，默认空实现
    func setupScrollView()

    /// 渲染滚动视图布局，setupSubviews之前调用，默认铺满
    func setupScrollLayout()
}

extension ScrollViewControllerProtocol where Self: UIViewController {
    /// 滚动视图，默认不显示滚动条
    public var scrollView: UIScrollView {
        if let result = fw.property(forName: "scrollView") as? UIScrollView {
            return result
        } else {
            let result = UIScrollView()
            result.showsVerticalScrollIndicator = false
            result.showsHorizontalScrollIndicator = false
            fw.setProperty(result, forName: "scrollView")
            return result
        }
    }

    /// 内容容器视图，自动撑开，子视图需要添加到此视图上
    public var contentView: UIView {
        if let result = fw.property(forName: "contentView") as? UIView {
            return result
        } else {
            let result = UIView()
            fw.setProperty(result, forName: "contentView")
            return result
        }
    }

    /// 渲染滚动视图，setupSubviews之前调用，默认空实现
    public func setupScrollView() {}

    /// 渲染滚动视图布局，setupSubviews之前调用，默认铺满
    public func setupScrollLayout() {
        scrollView.fw.pinEdges(autoScale: false)
    }
}

// MARK: - ViewControllerManager+ScrollViewControllerProtocol
extension ViewControllerManager {
    @MainActor func scrollViewControllerViewDidLoad(_ viewController: UIViewController) {
        guard let viewController = viewController as? UIViewController & ScrollViewControllerProtocol else { return }

        let scrollView = viewController.scrollView
        if let popupController = viewController as? PopupViewControllerProtocol {
            popupController.popupView.addSubview(scrollView)
        } else {
            viewController.view.addSubview(scrollView)
        }

        let contentView = viewController.contentView
        scrollView.addSubview(contentView)
        contentView.fw.pinEdges(autoScale: false)

        hookScrollViewController?(viewController)

        viewController.setupScrollView()
        viewController.setupScrollLayout()
        scrollView.setNeedsLayout()
        scrollView.layoutIfNeeded()
    }
}
